app/src/main/java/org/mozilla/focus/activity/MainActivity.kt

```diff
@@ -0,0 +1,300 @@
+
+    private fun showUrlInputScreen() {
+        val fragmentManager = supportFragmentManager
+        val browserFragment = fragmentManager.findFragmentByTag(BrowserFragment.FRAGMENT_TAG) as BrowserFragment?
+
+        val isShowingBrowser = browserFragment != null
+
+        if (isShowingBrowser) {
+            ViewUtils.showBrandedSnackbar(findViewById(android.R.id.content),
+                    R.string.feedback_erase,
+                    resources.getInteger(R.integer.erase_snackbar_delay))
+        }
+
+        // We add the url input fragment to the layout if it doesn't exist yet.
+        val transaction = fragmentManager
+                .beginTransaction()
+
+        // We only want to play the animation if a browser fragment is added and resumed.
+        // If it is not resumed then the application is currently in the process of resuming
+        // and the session was removed while the app was in the background (e.g. via the
+        // notification). In this case we do not want to show the content and remove the
+        // browser fragment immediately.
+        val shouldAnimate = isShowingBrowser && browserFragment!!.isResumed
+
+        if (shouldAnimate) {
+            transaction.setCustomAnimations(0, R.anim.erase_animation)
+        }
+
+        transaction
+                .replace(R.id.container, UrlInputFragment.createWithoutSession(), UrlInputFragment.FRAGMENT_TAG)
+                .commit()
+    }
+
+    private fun showFirstrun() {
+        supportFragmentManager
+                .beginTransaction()
+                .add(R.id.container, FirstrunFragment.create(), FirstrunFragment.FRAGMENT_TAG)
+                .commit()
+    }
+
+    protected fun showBrowserScreenForCurrentSession() {
+        val currentSession = currentSessionForActivity
+        val fragmentManager = supportFragmentManager
+
+        val fragment = fragmentManager.findFragmentByTag(BrowserFragment.FRAGMENT_TAG) as BrowserFragment?
+        if (fragment != null && fragment.session.isSameAs(currentSession)) {
+            // There's already a BrowserFragment displaying this session.
+            return
+        }
+
+        fragmentManager
+                .beginTransaction()
+                .replace(R.id.container, BrowserFragment.createForSession(currentSession), BrowserFragment.FRAGMENT_TAG)
+                .commit()
+    }
+
+    override fun onCreateView(name: String, context: Context, attrs: AttributeSet): View? {
+        return if (name == IWebView::class.java.name) {
+            // Inject our implementation of IWebView from the WebViewProvider.
+            WebViewProvider.create(this, attrs)
+        } else super.onCreateView(name, context, attrs)
+    }
+
+    override fun onBackPressed() {
+        val fragmentManager = supportFragmentManager
+
+        val sessionsSheetFragment = fragmentManager.findFragmentByTag(
+            SessionsSheetFragment.FRAGMENT_TAG) as SessionsSheetFragment?
+        if (sessionsSheetFragment != null &&
```

Comment:
We use this pattern three times in a row, reusing nearly identical code. It seems we could take advantage of Kotlin's lambdas and non-local return at label syntax to simplify this.